[AMDGPU] si-peephole-sdwa: Fix cndmask vcc use for wave32

frederik-h opened the pull request:
Before V_CNDMASK_B32_e64 gets converted to SDWA form, a conversion to V_CNDMASK_B32_e32 occurs.
The vcc use of this instruction must be fixed into a vcc_lo use for wave32. This fix only happens after the final
conversion to the SDWA form. This led to a compiler error in situations where the conversion to SDWA aborts.

Make sure that the vcc-fix gets applied even if the SDWA conversion is not completed.

frederik-h committed:
[AMDGPU] Add tests that demonstrates si-peephole-sdwa failure on V_CNDMASK

frederik-h committed:
[AMDGPU] si-peephole-sdwa: Fix cndmask vcc use for wave32

frederik-h committed:
Clean up test

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+
+
+; CHECK-NOT: {{.*}}V_CNDMASK_B32_e32{{.*}}$vcc
+; CHECK-NOT: {{.*}}Bad machine code: Virtual register defs don't dominate all uses
```
-NOT checks are next to useless, particularly for the error message. Just rely on the test not failing with the verifier

frederik-h replied:
I don't think they are useless. They serve as documentation - perhaps a comment would be just as good for this purpose - and they can improve the error message that you will see in case of a regression.

arsenm replied:
Can you also add this case to the mir test

frederik-h replied:
I think I cannot add it to the existing test file for "wave32" which uses "--run-pass=si-peephole-sdwa". The mir parser cannot read the mir that llc outputs with "--stop-before=si-peephole-sdwa" because it contains instructions with implicit "vcc_lo" operands. The parser insists on those being "vcc" instead because it does not take the wave32/wave64 distinction into account while checking the correctness of the implicit operands.

frederik-h replied:
Added it now after manually editing the mir output.

arsenm replied:
These are useless and should be removed. It's broken as soon as any detail of the output changes, and it is not the point of the test. This is not a verifier test, at most comment it's avoiding a verifier failure

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+  %extractelement = extractelement <5 x i8> %phi, i64 0
+  store i8 %extractelement, ptr addrspace(1) null, align 1
+  ret void
+}
```
```suggestion
}

```

Missing end of file newline

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1105,6 +1105,7 @@ void SIPeepholeSDWA::convertVcndmaskToVOP2(MachineInstr &MI,
                        .add(*TII->getNamedOperand(MI, AMDGPU::OpName::src0))
                        .add(*TII->getNamedOperand(MI, AMDGPU::OpName::src1))
                        .setMIFlags(MI.getFlags());
+  TII->fixImplicitOperands(*Converted);
```
It just occurred to me that this shouldn't be introducing the VOP2 form. It should directly go to SDWA or do nothing

frederik-h replied:
Right, but I think the same problem exists for the other instructions that are handled in a similar way, i.e. `V_SUB_CO_U32_e64` and `V_ADD_CO_U32_e64`. It's not necessarily wrong, but sub-optimal. I can try to get rid of this, but I suggest to bring in this bug fix first.

frederik-h committed:
Update llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Add mir test

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+; RUN: llc %s -march=amdgcn -mcpu=gfx1030 -o - 2>&1 | FileCheck %s
```
```suggestion
; RUN: llc -march=amdgcn -mcpu=gfx1030 < %s | FileCheck %s
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+
+; CHECK-NOT: {{.*}}V_CNDMASK_B32_e32{{.*}}$vcc
+; CHECK-NOT: {{.*}}Bad machine code: Virtual register defs don't dominate all uses
+; CHECK: {{.*}}v_cndmask_b32_e32{{.*}}vcc_lo
```
Not enough context, might as well generate checks

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+; CHECK-NOT: {{.*}}Bad machine code: Virtual register defs don't dominate all uses
+; CHECK: {{.*}}v_cndmask_b32_e32{{.*}}vcc_lo
+
+define amdgpu_kernel void @quux(i32 %arg, i1 %arg1, i1 %arg2) {
```
This can probably be reduced a lot, and can probably stop being a kernel

frederik-h replied:
This went trough llvm-reduce. I cannot reduce it signficantly by hand either. In particular, the control flow seems to be necessary. It doesn't need to be a kernel.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-wave32.mir:
```diff
@@ -230,3 +230,92 @@ body:             |
+    successors: %bb.1(0x40000000), %bb.2(0x40000000)
+    liveins: $vgpr0, $sgpr8_sgpr9
+
+    %0:sgpr_64(p4) = COPY $sgpr8_sgpr9
+    %1:vgpr_32(s32) = COPY $vgpr0
+    %2:sreg_64_xexec = S_LOAD_DWORDX2_IMM %0(p4), 0, 0
```
```suggestion
    %0:sgpr_64 = COPY $sgpr8_sgpr9
    %1:vgpr_32 = COPY $vgpr0
    %2:sreg_64_xexec = S_LOAD_DWORDX2_IMM %0, 0, 0
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-wave32.mir:
```diff
@@ -230,3 +230,92 @@ body:             |
+    %11:vgpr_32 = V_AND_B32_e64 %10, 255, implicit $exec
+    %12:vgpr_32 = V_AND_B32_e64 65535, killed %11, implicit $exec
+    S_CMP_EQ_U32 %6.sub0, 0, implicit-def $scc
+    %13:sreg_32_xm0_xexec = S_CSELECT_B32 -1, 0, implicit $scc
+    %14:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, killed %12, %13, implicit $exec
```
You should only need the minimal context around the v_cndmask_b32_e64, not the rest of the instructions. Definitely shouldn't need control flow

frederik-h replied:
I wasn't able to reduce the test case further. In particular, removing any of the branching removes the error.

frederik-h committed:
Update llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-wave32.mir

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Simplify and correct ll test

- Must use -mtriple to reproduce the bug on the unfixed branch
- Function does not need to be a kernel

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+; RUN: llc -march=amdgcn -mcpu=gfx1030 < %s | FileCheck %s
```
I only see the error with -mtriple=amdgcn-amd-amdhsa

frederik-h replied:
Correct. I already fixed that locally, too. Still trying to reduce the mir test case ...

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-fail.ll:
```diff
@@ -0,0 +1,40 @@
+; RUN: llc -march=amdgcn -mcpu=gfx1030 < %s | FileCheck %s
```
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 < %s | FileCheck %s
```

frederik-h committed:
Update test expectations

frederik-h committed:
Update mir test

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into peephole-sdwa-cndmask-fix-vcc
